### PR TITLE
PR: Research visual cryptography #20 

### PR DESCRIPTION
**Pull request by tchojnacki**

### CODE REVIEW

File: code/projects/steg/src/lib/src/vc.rs [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    FloydSteinberg = "floyd-steinberg",
+}
+
+fn closest_palette_color(v: u8) -> u8 {
+    if v < u8::MAX / 2 {
+        BLACK
+    } else {
+        WHITE
+    }
+}
```
Comment (mlodybercik) [inline code/projects/steg/src/lib/src/vc.rs:30]: 1. Thresholding could be done differently than just dividing the scale in half and quantizing to `MAX` or `MIN`.
2. nit: *Do we really need to make constants for `BLACK` and `WHITE`? Isn't it kinda obvious?*
If we need those two constants, why don't we create  `const HALFPOINT_THRESHOLD = u8::MAX / 2`? Maybe somebody does not understand what this means and needs a name for it.

File: code/projects/steg/src/lib/src/vc.rs [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        data[i + 2] = v;
+    }
+    create_image_data(&data, w as u32)
+}
+
+fn desaturate(image: &ImageData, grayscale: VcGrayscale) -> ImageData {
+    image.map_pixels(|(r, g, b)| {
+        let v = match grayscale {
+            VcGrayscale::Average => 0.333 * r as f64 + 0.333 * g as f64 + 0.333 * b as f64,
+            VcGrayscale::Luminosity => 0.299 * r as f64 + 0.587 * g as f64 + 0.114 * b as f64,
```
Comment (mlodybercik) [inline code/projects/steg/src/lib/src/vc.rs:50]: Average: Adding `r + g + b` and then dividing them in 3 could be better. We are loosing a little bit of precision for every 0.333 division.
Luminosity: Constants for `YUV` luminosity could be useful.

File: code/projects/steg/src/lib/src/vc.rs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+fn dither_random(image: &ImageData, seed: u64) -> ImageData {
+    let mut rng = StdRand::seed(seed);
+    image.map_pixels(|(v, _, _)| {
+        let v = if v < rng.next_u16() as u8 {
+            BLACK
+        } else {
+            WHITE
+        };
+        (v, v, v)
+    })
```
Comment (mlodybercik) [inline code/projects/steg/src/lib/src/vc.rs:70]: Same as previous, thresholding could be done as some value `x`: `0 < x < 255`.

**Review by jakubzehner — APPROVED** [review]
inbound4223414638596312064

File: code/projects/steg/src/lib/src/vc.rs (hunk 2 above)
Comment (tchojnacki) [reply code/projects/steg/src/lib/src/vc.rs:50]: Fixed locally, I'll push it to the remote soon.

**Comment by tchojnacki** [comment]
@mlodybercik comments resolved, merging

**Pull request merged**